httpreplay/cobweb.py

```diff
@@ -173,5 +187,214 @@ def handle(self, s, ts, protocol, sent, recv):
+        "rcpt to": "mail_to"
+    }
+
+    # Min to max SMTP response codes
+    _min = 100
+    _max = 600
+
+    def handle(self, s, ts, protocol, sent, recv):
+        if protocol != "tcp":
+            self.parent.handle(s, ts, protocol, sent, recv)
+            return
+
+        if self.stream is None:
+            self.stream = self.parent.tcp.streams[s]
+
+        self.parse_request(sent)
+        self.parse_reply(recv)
+
+        if self.stream.state in ["conn_finish", "conn_closed"]:
+            self.parent.handle(s, ts, "smtp", self.request, self.reply)
+
+    def init(self, *args, **kwargs):
+        self.request = SmtpRequest()
+        self.reply = SmtpReply()
+
+        # Current smtp command from client
+        self.command = None
+
+        # Last smtp server response code
+        self.last_rescode = 0
+
+        # Last smtp server response message
+        self.last_resmess = ""
+        self.stream = None
+
+    def _split(self, data, spliton="\r\n", always_list=False, maxsplit=-1):
+        splitdata = filter(None, data.split(spliton, maxsplit))
+        if len(splitdata) < 1:
+            return None
+        elif len(splitdata) == 1 and not always_list:
+            return splitdata[0]
+        else:
+            return splitdata
+
+    def _get_rescode(self, mes):
+        """"
+        Get the smtp server response code from a server reply
+        Returns None if no valid code or no code exists in the message
+        """
+        if len(mes) < 3:
+            return None
+
+        code = 0
+        try:
+            code = int(mes[:3])
+        except ValueError:
+            return None
+        if code >= self._min and code <= self._max:
+            return code
+
+    def get_command(self, request):
+        """
+        Get the smtp command from the request
+        Returns a used command. Returns None if command not valid
+        or no command exists in the message
+        """
+        command = None
+        res = request.lower()
+        for c in self._commands:
+            if res.startswith(c):
+                command = c
+                break
+
+        return command
+
+    def parse_request(self, request):
+        """"
+        Parses the requests sent by an smtp client. Values are stored in
+        a SmtpRequest object
+        """
+        self.request.raw.append(request)
+        self.command = self.get_command(request)
+
+        if self.command is not None:
+            r = re.compile(re.escape(self.command), re.IGNORECASE)
+            self.data = r.sub("", request)
+        else:
+            self.data = request
+
+        if self.command in self._command_field:
+            field = self._command_field[self.command]
+
+            if self.command in ["mail from", "rcpt to"]:
+                setattr(self.request, field,
+                        re.findall('<(.*?)>', self.data, re.DOTALL))
+            else:
+                setattr(self.request, field, self._split(self.data))
```

Comment:
Although `setattr()` may be useful, it'd probably be better (in terms of readability & simplicity) to create a dictionary mapping for each *implemented* command so that each SMTP command has its own simple handler. To get a better understanding of my idea, it is also used in Cuckoo as follows: https://github.com/cuckoosandbox/cuckoo/blob/package/cuckoo/processing/platform/windows.py#L198-L199
As you can see it attempts to see if there's a method called, e.g., `_api_pdf_escape`, and if it does, execute that handler. In this case you could do something like `_handle_mail_from` & `_handle_rcpt_to` etc.